### ISSUE TITLE
Support {"msg_type":"shutdown_request"} and “quit()” for “ipython console”

### ISSUE THREAD

**Issue by flying-sheep** [IRkernel/IRkernel#153]
start a R console with IPython frontend via `ipython console --kernel=ir`

pressing <kbd>C-d</kbd> will apparently send a `{"msg_type":"shutdown_request"}` message to the kernel. we should support those.

other methods to cleanly exit R like calling `quit()` should also work, but instead yield “kernel died, restart ([y]/n)?”

**Comment by takluyver** [comment]
There is code to support `shutdown_request` -  https://github.com/IRkernel/IRkernel/blob/0.4/R/kernel.r#L209

We should do the right thing when quit() is called, though. It's a bit convoluted: the kernel doesn't shut itself down, but the execute_reply it sends to the frontend includes an ask_exit [payload](http://ipython.org/ipython-doc/3/development/messaging.html#payloads). That will cause the frontend to issue a shutdown_request, and shut itself down as well.

**Comment by flying-sheep** [comment]
hmm, but something doesn’t work:

```console
$ ipython console --kernel=ir
IPython Console 3.1.0

In [1]: <C-d>
Do you really want to exit ([y]/n)? y
[1] "Got unhandled msg_type:" "shutdown_request"    
```

**Comment by takluyver** [comment]
Ah, I bet the shutdown_request is coming over the shell channel instead of the control channel. Technically, the same set of messages can go over either channel. I guess I was testing it on the Qt console, and there shutdown_request actually goes over the control channel.

**Comment by flying-sheep** [comment]
jup, if it can come via both, we need to handle it in both.

**Comment by flying-sheep** [comment]
ah, `quit()` is still not there

**Comment by JanSchulz** [comment]
Yep, also just found out that `quit()` is not the way to exit a kernel...

Maybe overwrite the default quit function to call whatever the kernel needs tell the kernel manager to quit?

**Comment by flying-sheep** [comment]
yes, but i think this is getting chaotic.

we need to structure everything to make clear how it works

* options we set
* functions and variables we insert into the namespace
* functions we override/replace

@takluyver what do we do apart from those?

**Comment by takluyver** [comment]
I don't think we currently inject anything into the user namespace. We inject one function of the `IRdisplay` namespace so that that can send data using the kernel.

In time we probably want to override `quit()` and `readline()` to use the kernel.

**Comment by flying-sheep** [comment]
we are [shading quit](https://github.com/IRkernel/IRkernel/blob/168357e4a918ee5913095f38454a0d5f6c8fc393/R/execution.r#L91) already.

but that breaks when clearing the global/user namespace, so maybe we should put it somewhere else. because yeah, it’s the user namespace, not ours.

**Comment by takluyver** [comment]
Aha, I missed that ;-).

What namespace is `quit()` normally in? Can we replace it there? Or can we shadow it from the package's namespace?

**Comment by flying-sheep** [comment]
> What namespace is `quit()` normally in?

you can easily check that by doing `environment(quit)`:

`<environment: namespace:base>`

> Can we replace it there? Or can we shadow it from the package's namespace?

i think both

**Comment by flying-sheep** [comment]
fixed ages ago in #196
